File: 2-Add Two Numbers/2Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        int value = 0;
        int ai, bi;
        ListNode dum(-1);
        ListNode *prev = &dum;
        for(ListNode *pa = l1, *pb = l2; 
            nullptr != pa || nullptr != pb;
            pa = (nullptr == pa ? nullptr : pa->next), pb = (nullptr == pb ? nullptr : pb->next), prev = prev-> next)
        {
            ai = (nullptr == pa ? 0 : pa->val);
            bi = (nullptr == pb ? 0 : pb->val);
            value = (ai + bi + carry) % 10;
            carry = (ai + bi + carry) / 10;
            prev->next = new ListNode(value);
        }
        if(0 < carry)
        {
            prev->next = new ListNode(carry);
        }
        return dum.next;
    }
};

File: source/2-Add Two Numbers/2Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        
        dum = ListNode(0)
        cur = dum
        car = 0

        while None != l1 or None != l2:
            sum = 0
            if l1 != None:
                sum += l1.val
                l1 = l1.next

            if l2 != None:
                sum += l2.val
                l2 = l2.next

            cur.next = ListNode((sum + car) % 10)
            car = (sum  + car) // 10
            cur = cur.next

        if 1 == car:
            cur.next = ListNode(car)

        return dum.next
        